BUG: Rare LCSIM null trackstate error

Running 2019 tritrig MC reconstruction on file 

`/sdf/group/hps/mc/4pt55GeV/tritrig/20um120nA/HPS_PhysicsRun2019-v2-FEE-Pass0/idealCond/tritrig_beam_readout_348.slcio`

crashes due to error pasted below... This error only occurred on 1/500 files reconstructed


```java.lang.NullPointerException: Cannot invoke "org.lcsim.event.TrackState.getLocation()" because "trackstate" is null
>...at org.lcsim.lcio.SIOTrack.write(SIOTrack.java:121)
>...at org.lcsim.lcio.SIOTrackBlockHandler.writeCollectionElement(SIOTrackBlockHandler.java:30)
>...at org.lcsim.lcio.AbstractBlockHandler.writeBlock(AbstractBlockHandler.java:55)
>...at org.lcsim.lcio.LCIOWriter.writeData(LCIOWriter.java:220)
>...at org.lcsim.lcio.LCIOWriter.write(LCIOWriter.java:254)
>...at org.lcsim.util.loop.LCIODriver.process(LCIODriver.java:114)
>...at org.lcsim.util.Driver.doProcess(Driver.java:261)
>...at org.lcsim.util.Driver.processChildren(Driver.java:271)
>...at org.lcsim.util.Driver.process(Driver.java:187)
>...at org.lcsim.util.DriverAdapter.recordSupplied(DriverAdapter.java:74)
>...at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:832)
>...at org.freehep.record.loop.DefaultRecordLoop.loop(DefaultRecordLoop.java:668)
>...at org.freehep.record.loop.DefaultRecordLoop.execute(DefaultRecordLoop.java:566)
>...at org.lcsim.util.loop.LCSimLoop.loop(LCSimLoop.java:152)
>...at org.lcsim.job.JobControlManager.run(JobControlManager.java:949)
>...at org.hps.job.JobManager.main(JobManager.java:31)
org.lcsim.job:SEVERE A fatal error occurred during the job
java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "org.lcsim.event.TrackState.getLocation()" because "trackstate" is null
>...at org.lcsim.util.loop.LCSimLoop.handleClientError(LCSimLoop.java:194)
>...at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:834)
>...at org.freehep.record.loop.DefaultRecordLoop.loop(DefaultRecordLoop.java:668)
>...at org.freehep.record.loop.DefaultRecordLoop.execute(DefaultRecordLoop.java:566)
>...at org.lcsim.util.loop.LCSimLoop.loop(LCSimLoop.java:152)
>...at org.lcsim.job.JobControlManager.run(JobControlManager.java:949)
>...at org.hps.job.JobManager.main(JobManager.java:31)
Caused by: java.lang.NullPointerException: Cannot invoke "org.lcsim.event.TrackState.getLocation()" because "trackstate" is null
>...at org.lcsim.lcio.SIOTrack.write(SIOTrack.java:121)
>...at org.lcsim.lcio.SIOTrackBlockHandler.writeCollectionElement(SIOTrackBlockHandler.java:30)
>...at org.lcsim.lcio.AbstractBlockHandler.writeBlock(AbstractBlockHandler.java:55)
>...at org.lcsim.lcio.LCIOWriter.writeData(LCIOWriter.java:220)
>...at org.lcsim.lcio.LCIOWriter.write(LCIOWriter.java:254)
>...at org.lcsim.util.loop.LCIODriver.process(LCIODriver.java:114)
>...at org.lcsim.util.Driver.doProcess(Driver.java:261)
>...at org.lcsim.util.Driver.processChildren(Driver.java:271)
>...at org.lcsim.util.Driver.process(Driver.java:187)
>...at org.lcsim.util.DriverAdapter.recordSupplied(DriverAdapter.java:74)
>...at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:832)
>...... 5 more
Exception in thread "main" java.lang.RuntimeException: Fatal error occurred during the job
>...at org.lcsim.job.JobControlManager.run(JobControlManager.java:972)
>...at org.hps.job.JobManager.main(JobManager.java:31)
Caused by: java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "org.lcsim.event.TrackState.getLocation()" because "trackstate" is null
>...at org.lcsim.util.loop.LCSimLoop.handleClientError(LCSimLoop.java:194)
>...at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:834)
>...at org.freehep.record.loop.DefaultRecordLoop.loop(DefaultRecordLoop.java:668)
>...at org.freehep.record.loop.DefaultRecordLoop.execute(DefaultRecordLoop.java:566)
>...at org.lcsim.util.loop.LCSimLoop.loop(LCSimLoop.java:152)
>...at org.lcsim.job.JobControlManager.run(JobControlManager.java:949)
>...... 1 more
Caused by: java.lang.NullPointerException: Cannot invoke "org.lcsim.event.TrackState.getLocation()" because "trackstate" is null
>...at org.lcsim.lcio.SIOTrack.write(SIOTrack.java:121)
>...at org.lcsim.lcio.SIOTrackBlockHandler.writeCollectionElement(SIOTrackBlockHandler.java:30)
>...at org.lcsim.lcio.AbstractBlockHandler.writeBlock(AbstractBlockHandler.java:55)
>...at org.lcsim.lcio.LCIOWriter.writeData(LCIOWriter.java:220)
>...at org.lcsim.lcio.LCIOWriter.write(LCIOWriter.java:254)
>...at org.lcsim.util.loop.LCIODriver.process(LCIODriver.java:114)
>...at org.lcsim.util.Driver.doProcess(Driver.java:261)
>...at org.lcsim.util.Driver.processChildren(Driver.java:271)
>...at org.lcsim.util.Driver.process(Driver.java:187)
>...at org.lcsim.util.DriverAdapter.recordSupplied(DriverAdapter.java:74)
>...at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:832)
>...... 5 more```

--- a/recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java
+++ b/recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java
@@ -328,6 +328,11 @@ public class TrackClusterMatcherMinDistance extends AbstractTrackClusterMatcher{
         else {
             TrackData trackdata = (TrackData) trackToData.from(track);
             TrackState ts_ecal = track.getTrackStates().get(track.getTrackStates().size()-1);
+            //If trackstate is null, upstream extrapolation error. Skip this
+            //Track
+            if(ts_ecal == null){
+                return null;
+            }
             double[] ts_ecalPos = ts_ecal.getReferencePoint();
             trackx = ts_ecalPos[1];
             tracky = ts_ecalPos[2];
@@ -425,6 +430,12 @@ public class TrackClusterMatcherMinDistance extends AbstractTrackClusterMatcher{
 
                 //Track positions depend on GBL or KF
                 List<Double> trackPos = this.getTrackPositionAtEcal(track);
+                //If trackPos null, Track extrapolation to Ecal failed. Skip.
+                if(trackPos == null){
+                    //
+                    trackClusterResidualsMap.put(track, null);
+                    continue;
+                }
                 double trackx = trackPos.get(0);
                 double tracky = trackPos.get(1);
                 double trackz = trackPos.get(2);
@@ -567,6 +578,10 @@ public class TrackClusterMatcherMinDistance extends AbstractTrackClusterMatcher{
             double trackPmag = new BasicHep3Vector(track.getTrackStates().get(0).getMomentum()).magnitude();
             //Track position
             List<Double> trackPos = this.getTrackPositionAtEcal(track);
+            //If trackPos is null, Track extrapolation to Ecal failed. Skip.
+            if(trackPos == null){
+                continue;
+            }
             double trackx = trackPos.get(0);
             double tracky = trackPos.get(1);
             double trackz = trackPos.get(2);
@@ -702,6 +717,10 @@ public class TrackClusterMatcherMinDistance extends AbstractTrackClusterMatcher{
             double smallestdr = 99999.0;
             Cluster smallestdrCluster = null;
             Map<Cluster, Double> cluster_dr_Map = trackClusterResidualsMap.get(track);
+            if(trackClusterResidualsMap == null){
+                Map.put(track, null);
+                continue;
+            }
             for(Cluster c : cluster_dr_Map.keySet()){
                 double dr = cluster_dr_Map.get(c);
                 if(dr < smallestdr){
@@ -728,6 +747,9 @@ public class TrackClusterMatcherMinDistance extends AbstractTrackClusterMatcher{
         boolean isTop;
 
         List<Double> trackPos = getTrackPositionAtEcal(track);
+        //If trackPos null, failed Track extrapolation. Skip Track
+        if(trackPos == null)
+            return;
         double trackx = trackPos.get(0);
         double tracky = trackPos.get(1);
         double trackz = trackPos.get(2);

